File: spring-security-in-action/ssia-ch07-ex3/src/test/java/org/example/ssiach07ex3/controller/TestControllerTest.java
package org.example.ssiach07ex3.controller;

import static org.junit.jupiter.api.Assertions.*;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.httpBasic;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;

@SpringBootTest
@AutoConfigureMockMvc(addFilters = true)
class TestControllerTest {
    @Autowired
    private MockMvc mvc;

    @Test
    void testProduct() throws Exception {
        mvc.perform(get("/product/1234").with(httpBasic("john", "12345")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").value("1234"));
    }

    @Test
    void testProduct2() throws Exception {
        mvc.perform(get("/product/asdf").with(httpBasic("john", "12345")))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$").value("asdf"));
    }
}

File: spring-security-in-action/ssia-ch5/src/test/kotlin/org/example/ssiach5/controller/TestControllerTest.kt
package org.example.ssiach5.controller

import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.httpBasic
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import kotlin.test.assertEquals

@SpringBootTest
@AutoConfigureMockMvc
class TestControllerTest {

    @Autowired
    private lateinit var mvc: MockMvc

    @Test
    fun helloTest() {
        mvc.perform(get("/hello").with(httpBasic("kwang", "1234")))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$").value("hello kwang!"))
            .andExpect(header().exists("Content-Length"))
            .andExpect(header().doesNotExist("message"))

    }

    @Test
    fun byeTest() {
        val result = mvc.perform(get("/bye").with(httpBasic("kwang", "1234")))
            .andExpect(status().isOk)
            .andReturn()

        val content = result.response.contentAsString
        assertEquals("bye kwang!", content)
    }

    @Test
    fun helloFailTest() {
        mvc.perform(get("/hello"))
            .andExpect(status().isUnauthorized)
            .andExpect(header().string("message", "Luke, I am your father!"))
    }
}